### PR TITLE
Implement kubeadm reset

### PR DESCRIPTION
**Pull request by luxas**
@kubernetes/sig-cluster-lifecycle

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34807)

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -150,7 +150,7 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, skipPreFlight
 	}
 
 	if !skipPreFlight {
-		fmt.Println("<cmd/init> Running pre-flight checks")
+		fmt.Println("Running pre-flight checks")
```
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/init.go]: Intentional?

File: cmd/kubeadm/app/preflight/checks.go [hunk 2]
```diff
@@ -191,10 +191,7 @@ func RunJoinNodeChecks() error {
 		IsRootCheck{root: true},
 		ServiceCheck{service: "docker"},
 		ServiceCheck{service: "kubelet"},
-		PortOpenCheck{port: 8080},
 		PortOpenCheck{port: 10250},
-		PortOpenCheck{port: 10251},
-		PortOpenCheck{port: 10252},
```
Comment (dgoodwin) [inline cmd/kubeadm/app/preflight/checks.go]: Did these turn out to not be necessary?

File: cmd/kubeadm/app/cmd/reset.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	} else {
+		fmt.Printf("no supported init system detected, skipping service checks for %s. kubeadm reset may not have full effect now.\n", serviceToStop)
+	}
+
+	fmt.Printf("Unmounting directories in /var/lib/kubelet...\n")
+	// Don't check for errors here, since umount will return a non-zero exit code if there is no directories to umount
+	exec.Command("sh", "-c", "cat /proc/mounts | awk '{print $2}' | grep '/var/lib/kubelet' | xargs umount").Run()
+
+	dirsToRemove := []string{"/var/lib/kubelet", "/var/lib/etcd", "/etc/kubernetes"}
+	fmt.Printf("Deleting the stateful directories: [%v]\n", dirsToRemove)
```
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/reset.go]: This shows up as: Deleting the stateful directories: [[/var/lib/kubelet /var/lib/etcd /etc/kubernetes]]

Can probably drop the [] in code.

File: cmd/kubeadm/app/cmd/reset.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	for _, dir := range dirsToRemove {
+		err := os.RemoveAll(dir)
+		if err != nil {
+			fmt.Errorf("failed to remove directory: [%v]", err)
+		}
+	}
+
+	fmt.Printf("Stopping all running docker containers...")
+	if err := exec.Command("sh", "-c", "docker ps -q | xargs docker kill").Run(); err != nil {
+		fmt.Printf("failed to stop the running containers")
```
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/reset.go]: Should we actually clean these up or just kill? My reset script I'd been clearing them out entirely.

Also should we only hit containers with a k8s prefix?

**Review by dgoodwin — COMMENTED** [review]
This worked great on CentOS, just a few questions and one very minor request for changes.

File: cmd/kubeadm/app/cmd/reset.go (hunk 4 above)
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/reset.go]: Thinking we may want to filter to k8s containers only, I could envision a developer deciding to init master on their workstation.

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Yep, I think we decided to drop it
At least it was dropped in join

File: cmd/kubeadm/app/cmd/reset.go (hunk 4 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: I thought about that, leads to more work, but is probably worth it

Not sure if we should remove them though... the user might want to read logs

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Drop it.

File: cmd/kubeadm/app/preflight/checks.go (hunk 2 above)
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: This is for the node, 10251 and 10252 for cm and scheduler and 8080 for unsecured apiserver

File: cmd/kubeadm/app/cmd/reset.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+		fmt.Println("Skipping pre-flight checks")
+	}
+
+	serviceToStop := "kubelet"
+	initSystem := initsystem.GetInitSystem()
+	if initSystem != nil {
+		fmt.Printf("Stopping the %s service...\n", serviceToStop)
+		initSystem.ServiceStop(serviceToStop)
+	} else {
+		fmt.Printf("no supported init system detected, skipping service checks for %s. kubeadm reset may not have full effect now.\n", serviceToStop)
```
Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: Why doesn't `initsystem.GetInitSystem()` return that string as an error?

File: cmd/kubeadm/app/cmd/reset.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	fmt.Printf("Unmounting directories in /var/lib/kubelet...\n")
+	// Don't check for errors here, since umount will return a non-zero exit code if there is no directories to umount
+	exec.Command("sh", "-c", "cat /proc/mounts | awk '{print $2}' | grep '/var/lib/kubelet' | xargs umount").Run()
```
Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: I think `find /var/lib/kubelet | xargs -n 1 findmnt -n -t tmpfs -o TARGET -T | uniq | xargs -r umount ` is cleaner.

Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: If you depend on `awk`, you don't have to use `grep`, i.e. `awk '$2 ~ /\/var\/lib\/kubelet { print $2 }' | xargs umount`, or even `awk '$2 ~ /\/var\/lib\/kubelet { system("umount "$2) }`, on the other hand you don't even have to shell out...

File: cmd/kubeadm/app/cmd/reset.go (hunk 4 above)
Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/reset.go]: I think we should use `docker rm --force --volumes` here.

> the user might want to read logs

Maybe, but more user will complain about disk space used up by those containers.

**Review by errordeveloper — CHANGES_REQUESTED** [review]

**Comment by luxas** [comment]
PTAL @pires @mikedanese @dgoodwin @errordeveloper

**Comment by luxas** [comment]
I'd like to get this lgtm'd ASAP, if @mikedanese could do it, it would be awesome! (Ilya is offline now)

**Review by dgoodwin — APPROVED** [review]
LGTM

**Comment by luxas** [comment]
Applying @dgoodwin's LGTM

File: cmd/kubeadm/app/cmd/reset.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	cmd.PersistentFlags().BoolVar(
+		&skipPreFlight, "skip-preflight-checks", false,
+		"skip preflight checks normally run before modifying the system",
+	)
+
+	return cmd
+}
+
+func RunReset(out io.Writer, cmd *cobra.Command, skipPreFlight bool) error {
+
```
Comment (pires) [inline cmd/kubeadm/app/cmd/reset.go]: Why the empty line?

File: cmd/kubeadm/app/cmd/reset.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+	if !skipPreFlight {
+		fmt.Println("Running pre-flight checks")
+		err := preflight.RunResetCheck()
+		if err != nil {
+			return err
```
Comment (pires) [inline cmd/kubeadm/app/cmd/reset.go]: This will break pre-flight checks. It should be `return &preflight.PreFlightError{Msg: err.Error()}}` instead.

**Review by pires — CHANGES_REQUESTED** [review]

**Comment by luxas** [comment]
Fixed @pires' comments, waiting for the build to be green before applying lgtm again

**Review by pires — APPROVED** [review]
LGTM

**Pull request merged**